Add new linter: dict-iter-missing-items (E1141)

yory8 opened the pull request:
<!--

Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

You can also read more about contributing to pylint in this document:
https://github.com/PyCQA/pylint/blob/master/doc/development_guide/contribute.rst#repository
-->

## Steps

- [X] Add yourself to CONTRIBUTORS if you are a new contributor.
- [X] Add a ChangeLog entry describing what your PR does.
- [X] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [X] Write a good description on what the PR does.

## Description

Add a new linter to check against forgotten calls to .items() when iterating through a dict in a for loop.

While the linter itself and its tests work fine standalone, I can't figure out how to integrate them in the repo, so **this pull request is broken** and waiting for help.

## Type of Changes
<!-- Leave the corresponding lines for the applicable type of change: -->
|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

## Related Issue

Closes #2761

PCManticore commented on pylint/checkers/typecheck.py:
```diff
@@ -356,6 +356,12 @@ def _missing_member_hint(owner, attrname, distance_threshold, max_choices):
         "(i.e. doesn't define __hash__ method).",
     ),
+    "E1141": (
+        "Dict is not iterable",
```
Let's add a bit more context here, since the dictionary is actually iterable, but it's going to raise an error if we don't call `items()` on it. So maybe something along the lines of `Unpacking a dictionary in iteration without calling .items()`?

yory8 replied:
fixed

PCManticore commented on pylint/test/unittest_checker_typecheck.py:
```diff
@@ -333,3 +333,69 @@ class AggregateCls:
         with self.assertNoMessages():
             self.checker.visit_call(call)
+
+    def test_dict_iter_missing_items(self):
```
We usually place functional tests in `pylint/test/functional`, where each Python file has a name of a check that its testing against. There is also a corresponding `<message>.txt` file which contains the output that we expect for that check.

PCManticore commented on pylint/checkers/typecheck.py:
```diff
@@ -1525,6 +1531,32 @@ def visit_subscript(self, node):
+        if not isinstance(node.target, astroid.node_classes.Tuple):
+            return
+
+        itr = node.iter
```
`iterable` maybe? it's more obvious on what it does that `itr`.

PCManticore commented on pylint/checkers/typecheck.py:
```diff
@@ -1525,6 +1531,32 @@ def visit_subscript(self, node):
+        itr = node.iter
+
+        # not method
+        if not isinstance(itr, astroid.node_classes.Call):
```
This suggests to me that it only looks for `Call` nodes as the iterable, is that the case? I think a simpler solution would be to:
- look for `Name`. This usually indicates that we tried to iterate a dictionary stashed in a variable
- Call`pylint.checkers.utils.safe_infer(iterable)` and check that the return value is a `Dict`
- If it is a Dict and if we have a tuple of two values in the `target`, then emit the error

This should catch most of the occurrences and it will also have less false positives.

yory8 commented:
I addressed the issues found in your review. I'm still having problems, though: the tests fail, I think because my "visit_for" check in typechecks.py is actually ignored, which seems weird.

yory8 commented:
So, apparently the patch actually is fine. It works when testing with tox, just not with pytest. :confused:

yory8 commented:
@PCManticore I'm having issues with formatting. On my system, black (same version as the one in travis) tells me there's nothing to format, whereas when running in travis it complains.
After this is solved, the patch should be ready for merging.

PCManticore commented on pylint/checkers/typecheck.py:
```diff
@@ -1525,6 +1531,25 @@ def visit_subscript(self, node):
+        if not isinstance(node.target, astroid.node_classes.Tuple):
+            # target is not a tuple
+            return
+        if not len(list(node.target.get_children())) == 2:
```
This can be `node.target.elts` instead of calling `get_children()`

PCManticore commented on pylint/checkers/typecheck.py:
```diff
@@ -1525,6 +1531,25 @@ def visit_subscript(self, node):
+            # it's not a bare variable
+            return
+
+        if not isinstance(safe_infer(iterable), astroid.node_classes.Dict):
```
You need to verify the output of `safe_infer` before using it, as it might return either `None` or `Uninferable` in case it can't infer the given object:
```
inferred = safe_infer(iterable)
if not inferred:
    return
if not isinstance(inferred, Dict):
    # emit the message
```

PCManticore commented on pylint/test/functional/dict_iter_missing_items.py:
```diff
@@ -0,0 +1,19 @@
+for k, v in d:  # [dict-iter-missing-items]
+    pass
+
+# Negative
```
Can you also add some tests with uninferable results? e.g.
```
from unknown import Uninferable

for k, v in Uninferable: # should not trigger
    pass
```

PCManticore commented:
@yory8 Are you using `black==18.6b4`? Can you try locally with `tox -e formatting`, do you still get errors with that?

yory8 commented:
About the review, I'll work on it later. About formatting, I used `tox -e formatting` locally and it says everything is fine, it's the one in Travis that complains about it, I don't know why :confused:

PCManticore commented:
@yory8 I don't know exactly why you are getting the formatting error. You might try with `tox -e formatting --recreate` to recreate the tox environment, which could bring a newer `black`. Regardless, the tests on AppVeyor are passing, I believe this is good to go. I'll merge it as is, followed by an update to `typecheck.py` to get the formatting error in place.

Thank you so much for your PR! 🎉

yory8 commented:
Great!
 Thank'you for your guidance.
